opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -35,42 +39,59 @@ impl<T: Number<T>> ValueMap<T> {
-                    // Recheck after acquiring write lock, in case another
-                    // thread has added the value.
-                    if let Some(value_to_update) = values.get(&attrs) {
-                        value_to_update.add(measurement);
-                        return;
-                    } else if is_under_cardinality_limit(values.len()) {
-                        let new_value = T::new_atomic_tracker();
-                        new_value.add(measurement);
-                        values.insert(attrs, new_value);
-                    } else if let Some(overflow_value) =
-                        values.get_mut(&STREAM_OVERFLOW_ATTRIBUTE_SET)
-                    {
-                        overflow_value.add(measurement);
-                        return;
-                    } else {
-                        let new_value = T::new_atomic_tracker();
-                        new_value.add(measurement);
-                        values.insert(STREAM_OVERFLOW_ATTRIBUTE_SET.clone(), new_value);
-                        global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow. Subsequent overflows to same metric until next collect will not be logged.".into()));
+                // Then try sorted order.
+                let sorted_attrs = AttributeSet::from(attrs).into_vec();
+                if let Some(value_to_update) = values.get(sorted_attrs.as_slice()) {
+                    value_to_update.add(measurement);
+                } else {
+                    // Give up the lock, before acquiring write lock.
+                    drop(values);
+                    if let Ok(mut values) = self.values.write() {
+                        // Recheck both incoming and sorted after acquiring
+                        // write lock, in case another thread has added the
+                        // value.
+                        if let Some(value_to_update) = values.get(attrs) {
+                            value_to_update.add(measurement);
+                        } else if let Some(value_to_update) = values.get(sorted_attrs.as_slice()) {
+                            value_to_update.add(measurement);
+                        } else if is_under_cardinality_limit(self.count.load(Ordering::SeqCst)) {
+                            let new_value = T::new_atomic_tracker();
+                            new_value.add(measurement);
+                            let new_value = Arc::new(new_value);
+
+                            // Insert original order
+                            values.insert(attrs.to_vec(), new_value.clone());
+
+                            // Insert sorted order
+                            values.insert(sorted_attrs, new_value);
+
+                            self.count.fetch_add(1, Ordering::SeqCst);
+                        } else if let Some(overflow_value) =
+                            values.get_mut(STREAM_OVERFLOW_ATTRIBUTES.as_slice())
```

Comment:
I think we could just use the `get` method here.

```suggestion
                            values.get(STREAM_OVERFLOW_ATTRIBUTES.as_slice())
```